ls1intum/Artemis
# [Exam] Add/Import modeling Exercises

### Checklist

### Motivation and Context
As part of the artemis exams, we want to also support modeling exercises, by creating new ones but also by having the ability to import modeling exercises. In addition to this, we want to import modeling exercises into courses. 


### Description
- [x] Create the import modal for Modeling Exercises
- [x] Add Pagable Search calls for the import modal 
- [x] Adapt the modeling-exercise-update.component.ts for import and exam mode
- [x] Adapt the modeling-exercise-update.component.html for import and exam mode
- [x] Add the import routes for modeling exercises for course and exam
- [x] Update the ModelingExerciseResolver to handle requests from exams
- [x] Create the server rest call 
- [x] Create the abstract class ExerciseImportService, let TextExerciseImportService extend ExerciseImportService
- [x] Create the ModelingExerciseImportService, extend ExerciseImportService
- [x] Create the database call required for the import service
- [x] Update the exerciseGroup row buttons to handle deletion of modeling exercises

### Steps for Testing
Firstly, make sure importing text exercises still works as expected for courses and exams. There was some code refactoring involved here, but no functionality change. 

To test the modeling exercises: 
**Course:**
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a modeling exercise in your course. Add every field possible including the example submission and example solution
4. Import this exercise into the course using the import modeling exercise button. 
5. Delete the original exercise. Everything should still work and be visible for the imported exercise.
----------------------------------------------
**Exam:**
1. Log in to Artemis
2. Navigate to Course Administration
3. Create an exam and set it up with two exercise groups, let it start soon.
4. Add one modeling exercise to one exercise group. 
5. Import a modeling exercise to the other exercise group. 
6. Start the exam.
7. Do some actions and save

### Screenshots
Course Import Modeling Exercise
![image](https://user-images.githubusercontent.com/15110960/86280849-df31bf00-bbdc-11ea-83a9-2c08591bc62d.png)

Import Modeling Exercise Modal
![image](https://user-images.githubusercontent.com/15110960/86281509-f329f080-bbdd-11ea-9b0b-bdfcafce04b6.png)

Exercise Group Modeling Exercise Actions
![image](https://user-images.githubusercontent.com/15110960/86280928-012b4180-bbdd-11ea-9e59-104bd61a9977.png)

Modeling Exercises in the exam: 
![image](https://user-images.githubusercontent.com/15110960/86280974-14d6a800-bbdd-11ea-8091-488059f18e9e.png)
